### DISCUSSION TITLE
瓦片图层内部实现架构优化

### DISCUSSION THREAD

**Discussion by lvisei** [antvis/L7#2333] [Ideas]
Ref 
- #2322 
-  [chore: tilelayer scene 模块初始化](https://github.com/antvis/L7/pull/2322/commits/c01e8b5286819e2d3f40067c4a58246b419424b7)

现在瓦片内部实现的依赖关系：

```
LineLayer -> core/BaseLayer -> plugins/LayerModelPlugin -> skip init Layer Model -> init TileLayer（scene 里面处理）  -> VectorTile -> layers/tile/core/BaseTileLayer -> LineLayer -> core/BaseLayer -> plugins/LayerModelPlugin -> init Layer
```

1. 实例化线瓦片图层，线瓦片数据跳过 Model 流程：

```
LineLayer -> core/BaseLayer -> plugins/LayerModelPlugin -> skip init Layer Model
```

2. layer 初始化完成，由 scene 实例化瓦片逻辑：

```
layer.tileLayer = new TileLayer(layer)  ->  VectorTile -> layers/tile/core/BaseTileLayer[] -> LineLayer -> core/BaseLayer -> plugins/LayerModelPlugin -> init Layer

```

对于瓦片，现在的 BaseLayer 其实是一个复合图层概念，但复合图层的概念没有抽离，导致现在的逻辑晦涩，对象关系依赖循环。L7 通过基本图形透出瓦片的概念，与 Mapbox API 风格一致，但 Mapbox 采用了瓦片坐标系架构，内置的渲染单位为瓦片单元，在概念上与 L7 的基础图层不对等，故不存在图层嵌套情况。要优化 L7 的瓦片图层代码逻辑：

- 如果支持瓦片坐标系架构，很多计算需要放到 CPU 里面执行，需要上 worker，内部数据结构以瓦片形式组织，架构变动很大。
- 就现有架构来看，可能还是需要 「抽象复合图层概念」 +「 瓦片图层改复合图层」方案。

**Comment by lvisei** [comment]
> https://github.com/antvis/L7/pull/2322#issuecomment-1978571811

多 Model 也是一个方案，单个 Model 实现一个瓦片的渲染，需要权衡用来当作瓦片图层实现，对整个 model 的管理的成本带来的性能问题，现在属性更新机制现在是做到图层上面的，与 model 是分离的，属性的更新会导致所有 model 的重新构建问题。

**Reply by lvisei** [reply]
基础 layer 这一层做好 属性管理 + model，多 model 来实现瓦片图层，如果还做对应的多 「属性管理」 机制，让最基本图层单元变复杂了很多，设计还是最小粒度，分层实现更加容易维护。

**Reply by lzxue** [reply]
现在的 Model 更新，确实没有做按需更新，数据更新了进行就进行 Model 重建，不过重建 Model 不涉及复杂数据运算，这块性能是可控的。

目前 Layer 的数据更新和三角网更新做到解耦了，颜色更新的时候，三角网不会重新计算，
